src/client/opamClient.ml

```diff
@@ -315,6 +315,63 @@ let upgrade_t ?strict_upgrade ?auto_install ?ask ?(check=false) ~all atoms t =
              (OpamPackage.names_of_packages t.compiler_packages)
          in
          let unopt = unopt -- base in
+         let get_formula pkg get =
+           OpamStd.Option.map
+             (fun opam ->
+                OpamFilter.filter_formula ~default:false
+                  (OpamFilter.deps_var_env ~build:true ~post:false
+                     ~test:true ~dev:false ~doc:false)
```

Comment:
why this choice ? My guess is that this should be
- `build:true`, `post:true` (these are the "base" dependencies as seen by the solver)
- `dev` should be determined from `OpamSwitchState.is_dev_package` IIRC
- `test` and `doc` should be determined from `!OpamStateConfig.r.build_{test,doc}`, and wether the package is a member of the `requested` set.

Most of this should be handled correctly already by `OpamPackageVar.all_depends` (plus the possible `version` variable)